### ISSUE TITLE
models trained on previous versions do not work correctly

### ISSUE THREAD

**Issue by krylea** [pytorch/pytorch#8656]
## Issue description

I have a sequence to sequence model trained with pytorch 0.3.0.post4. It gave reasonable outputs when queried at the time, and I have since downloaded the model onto my computer (which uses 0.3.1.post2) and still observe reasonable responses from the model.
The server this model was trained on was recently upgraded to pytorch 0.4, and since then the model produces nonsensical results when queried. I need to use this trained model's weights as the initial weights for another network (which I am training on the same server) and this new network also produces nonsensical results when it is trained and queried.

I am not sure how to attach a code sample as the model I am training is quite complicated, and the code is several thousand lines long. There is no one particular place the code is failing, the outputs simply do not make sense.

## System Info
Server:
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB
GPU 2: Tesla P100-PCIE-16GB
GPU 3: Tesla P100-PCIE-16GB

Nvidia driver version: 384.111
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] msgpack-numpy (0.4.1)
[pip3] numpy (1.13.3)
[pip3] torch (0.4.0)
[pip3] torchtext (0.2.3)
[pip3] torchvision (0.1.9)
[conda] Could not collect

Home Computer:
PyTorch version: 0.3.1.post2
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.2
GCC version: Could not collect
CMake version: version 3.6.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.3)
[pip3] numpydoc (0.6.0)
[pip3] pytorchviz (0.0.1)
[pip3] torch (0.3.1.post2)
[pip3] torchvision (0.2.0)
[conda] pytorch                   0.3.1           py36_cuda0.0_cudnn0.0_2    pytorch
[conda] pytorchviz                0.0.1                     <pip>
[conda] torchvision               0.2.0            py36hf5eb7ec_1    pytorch

**Comment by SsnL** [comment]
Maybe a list of type of layers used is a good start for us to debug. :)

**Comment by krylea** [comment]
The model is a standard naive sequence to sequence language model with Luong attention, using 1-layer GRUs as the base RNNs. Inputs are passed into an embedding layer, then an "encoder" RNN (1-layer bidirectional GRU). The output hidden state from this is then used as the initial hidden state for a "decoder" RNN (1-layer unidirectional GRU). The decoder is given a predefined token at each decoding step, and at each step the output of the decoder is concatenated with a context vector computed by Luong attention over the encoder outputs (weighted sum of bidirectional outputs at each encoding step, with weights determined by the dot product between the encoder's output at that step and the output of the decoder RNN), then passed through a linear layer with tanh activation, and finally passed into a last linear layer and softmaxed to produce a probability distribution over the vocabulary for selecting the next token.
Training is done in a batched fashion, with masking, using a cross-entropy loss.

**Comment by ezyang** [comment]
Even if the model is complicated, if you can post it in some sort of self-contained form, that would still be helpful.

If not, here are some things to try:

1. Get the model running on master, and see if it gives nonsense or not
2. Construct a test case that can distinguish between "correct" and "nonsense" outputs, and then bisect commits between 0.3 and 0.4 which cause problems
3. Dump the intermediate results of the model, and determine which layer is causing the divergence.

**Comment by krylea** [comment]
Hmm, it looks like it might actually just be some sort of issue with how the outputs are formatted now that is leading to unexpected results.
